lib_async/kafka_async.mli

```diff
@@ -0,0 +1,28 @@
+open Async
+
+type producer
+
+type consumer
+
+type 'a response = ('a, Kafka.error * string) result
+
+val produce :
+  producer ->
+  Kafka.topic ->
+  Kafka.partition ->
+  ?key:string ->
+  string ->
+  unit response Deferred.t
+
+val new_producer : (string * string) list -> producer response
+
+val new_consumer : (string * string) list -> consumer response
+
+val new_topic :
+  producer -> string -> (string * string) list -> Kafka.topic response
```

Comment:
Maybe a type `handler`, wrapping either a `producer` or a `consumer`, might be useful here.

We can also discuss the sync and Lwt interfaces. It can be helpful to distinguish producers and consumers rather than having only handlers.

Reply:
I agree. In the beginning I had just one handler type, but looking at the way handlers for consumers & producers are created in rdkafka they are distinct (either `RD_KAFKA_PRODUCER` or `RD_KAFKA_CONSUMER`) and I don't think e.g. using a producer for consuming would work. The C code just kinda ignores it and just calls both of them `rd_kafka_t`, thus allowing to misuse the API.

So I thought it would be nice to avoid this potential pitfall (given that we probably want a program that compiles to be somewhat correct) to use a phantom type to distinguish them on the OCaml side. But the more I worked on it, the more the contents of the OCaml types diverged to a point where I thought that making them distinct, unrelated types is better. This helped me to avoid making the `producer` and `consumer` types polymorphic but it unfortunately has lead to a bit of duplication of code. That can probably be cleaned up by creating a common subset of these types, a e.g. a `connection` type which would contain the `start`, `stop` and `handle` keys.

It also unfortunately means that producers and consumers have to be closed using different functions, but I judged the added type safety is worth it.

Reply:
I agree: the benefits of typed handlers worth the value, even if there is some duplicate code.